configapi/api_inventory.go

```diff
@@ -0,0 +1,231 @@
+        MsgMethod:   configmodels.Post_op,
+        GnbName: gnbName,
+        Gnb: &procReq,
+    }
+    configChannel <- &msg
+    configLog.Infof("Successfully added gNB [%v] to config channel.", gnbName)
+    return nil
+}
+
+func handleDeleteGnb(c *gin.Context) error {
+    var gnbName string
+    var exists bool
+    if gnbName, exists = c.Params.Get("gnb-name"); !exists {
+        errorMessage := "Delete gNB request is missing gnb-name"
+        configLog.Errorf(errorMessage)
+        return fmt.Errorf(errorMessage)
+    }
+    configLog.Infof("Received delete gNB %v request", gnbName)
+    msg := configmodels.ConfigMessage{
+        MsgType:   configmodels.Inventory,
+        MsgMethod: configmodels.Delete_op,
+        GnbName:   gnbName,
+    }
+    configChannel <- &msg
+    configLog.Infof("Successfully added gNB [%v] with delete_op to config channel.", gnbName)
+    return nil
+}
+
+func GetUpfs(c *gin.Context) {
+    setInventoryCorsHeader(c)
+    logger.WebUILog.Infoln("Get all UPFs")
+
+    var upfs []*configmodels.Upf
+    upfs = make([]*configmodels.Upf, 0)
+    rawUpfs, errGetMany := dbadapter.CommonDBClient.RestfulAPIGetMany(upfDataColl, bson.M{})
+    if errGetMany != nil {
+        logger.DbLog.Errorln(errGetMany)
+        c.JSON(http.StatusInternalServerError, upfs)
+    }
+
+    for _, rawUpf := range rawUpfs {
+        var upfData configmodels.Upf
+        err := json.Unmarshal(mapToByte(rawUpf), &upfData)
+        if err != nil {
+            logger.DbLog.Errorf("Could not unmarshall UPF %v", rawUpf)
+        }
+        upfs = append(upfs, &upfData)
+    }
+    c.JSON(http.StatusOK, upfs)
+}
+
+func PostUpf(c *gin.Context) {
+    setInventoryCorsHeader(c)
+    if err := handlePostUpf(c); err == nil {
+        c.JSON(http.StatusOK, gin.H{})
+    } else {
+        c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+    }
+}
+
+func DeleteUpf(c *gin.Context) {
+    setInventoryCorsHeader(c)
+    if err := handleDeleteUpf(c); err == nil {
+        c.JSON(http.StatusOK, gin.H{})
+    } else {
+        c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+    }
+}
+
+func handlePostUpf(c *gin.Context) error {
+    var upfHostname string
+    var exists bool
+    if upfHostname, exists = c.Params.Get("upf-hostname"); !exists {
+        errorMessage := "Post UPF request is missing upf-hostname"
+        configLog.Errorf(errorMessage)
+        return errors.New(errorMessage)
+    }
+    configLog.Infof("Received UPF %v", upfHostname)
+    var err error
+    var newUpf configmodels.Upf
+
+    allowHeader := strings.Split(c.GetHeader("Content-Type"), ";")
+    switch allowHeader[0] {
+    case "application/json":
+        err = c.ShouldBindJSON(&newUpf)
+    }
+    if err != nil {
+        configLog.Errorf("err %v", err)
+        return fmt.Errorf("Failed to create UPF %v: %w", upfHostname, err)
+    }
+    if newUpf.Port == "" {
+        errorMessage := "Post UPF request body is missing port"
```

Comment:
Would like to know how UPF port is used and what it should be.

Reply:
The UPF port is specified in the parameters when creating a new network slice.
```
    "upf": {
      "upf-name": "upf-external.testpcf.svc.cluster.local",
      "upf-port": "8805"
    }
```
We are keeping the parameter requirements consistent with what is asked for network slice creation, even if we use 8805.